File: Solutions/242-valid-anagram.py
class Solution:
    def isAnagram(self, s: str, t: str) -> bool:
        if len(s) != len(t):
            return False

        ds, dt = {}, {}

        for i in range(len(s)):
            ds[s[i]] = 1 + ds.get(s[i], 0)
            dt[t[i]] = 1 + dt.get(t[i], 0)
        return ds == dt
        
        #Time Complexity : O(N)
        #Space Complexity : O(N)

File: Solutions/Java/242-valid-anagram.java
/*
 * Time complexity: O(n)
 * Space complexity: O(1), since input strings
 * consist of lowercase English letters only
 *
 */

class Solution {
    public boolean isAnagram(String s, String t) {
        HashMap<Character, Integer> charsCount = new HashMap<>();

        for(char letter : s.toCharArray()) {
            charsCount.put(letter, charsCount.getOrDefault(letter, 0) + 1);
        }

        for(char letter : t.toCharArray()) {
            if(!charsCount.containsKey(letter)) {
                return false;
            } else {
                int count = charsCount.get(letter);
                if(count == 0) {
                    return false;
                } else {
                    charsCount.put(letter, count-1);
                }
            }
        }

        for(Character letter : charsCount.keySet()) {
            if(charsCount.get(letter) != 0) {
                return false;
            }
        }

        return true;
    }
}